drgn_tools/meminfo.py

```diff
@@ -0,0 +1,678 @@
+    to the kernel's memory subsystem. Results are cached in a dictionary.
+
+    :param prog: drgn program
+    :returns: A dictionary that contains relevant systemwise parameters
+    """
+    cache = prog.cache.setdefault("drgn_tools.meminfo", {})
+    mm_consts = cache.get("mm_constants", {})
+    if not mm_consts:
+        mm_consts = {}
+
+        # Determine max numbers of NUMA nodes and memory zones in a node.
+        # In mm/mempolicy.c, ```preferred_node_policy``` is defined as the following:
+        # "static struct mempolicy preferred_node_policy[MAX_NUMNODES];"
+        # Also, ```MAX_NUMNODES``` equals to (1 << NODES_SHIFT).
+        mm_consts["MAX_NUM_NODES"] = len(prog["preferred_node_policy"])
+        mm_consts["NODES_SHIFT"] = int(math.log2(mm_consts["MAX_NUM_NODES"]))
+        mm_consts["MAX_NUM_ZONES"] = prog["__MAX_NR_ZONES"].value_()
+
+        # Determine page size
+        mm_consts["PAGE_SIZE"] = prog.constant("PAGE_SIZE").value_()
+        mm_consts["PAGE_SHIFT"] = prog.constant("PAGE_SHIFT").value_()
+
+        # Determine architecture-specific parameters
+        if prog.platform.arch == drgn.Architecture.X86_64:
+            # arch/x86/include/asm/pgtable_64_types.h
+            _pmd_shift = 21
+            if "vmalloc_base" in prog:
+                _vmalloc_size_tb = 32
+                mm_consts["VMALLOC_START"] = prog["vmalloc_base"].value_()
+                mm_consts["VMALLOC_END"] = (
+                    mm_consts["VMALLOC_START"] + (_vmalloc_size_tb << 40) - 1
+                )
+            else:
+                mm_consts["VMALLOC_START"] = 0xFFFFC90000000000
+                mm_consts["VMALLOC_END"] = 0xFFFFE8FFFFFFFFFF
+        elif prog.platform.arch == drgn.Architecture.AARCH64:
+            # arch/arm64/include/asm/pgtable-hwdef.h
+            n = 2
+            _pmd_shift = (mm_consts["PAGE_SHIFT"] - 3) * (4 - (n)) + 3
+            n = 1
+            _pgdir_shift = (mm_consts["PAGE_SHIFT"] - 3) * (4 - (n)) + 3
+            # include/asm-generic/pgtable-nopud.h
+            _pud_shift = _pgdir_shift
+            _pud_size = 1 << _pud_shift
+
+            # arch/arm64/include/asm/pgtable.h
+            _sz_64k = 0x00010000
+            _sz_128m = 0x08000000
+            _va_bits = 48
+            _va_start = 0xFFFFFFFFFFFFFFFF - (1 << _va_bits) + 1
+            _page_offset = 0xFFFFFFFFFFFFFFFF - (1 << (_va_bits - 1)) + 1
+            _struct_page_max_shift = 6
+            _kasan_shadow_size = 0
+            _modules_vaddr = _va_start + _kasan_shadow_size
+            _modules_vsize = _sz_128m
+            _modules_end = _modules_vaddr + _modules_vsize
+            _vmemmap_size = 1 << (
+                _va_bits - mm_consts["PAGE_SHIFT"] - 1 + _struct_page_max_shift
+            )
+
+            mm_consts["VMALLOC_START"] = _modules_end
+            mm_consts["VMALLOC_END"] = (
+                _page_offset - _pud_size - _vmemmap_size - _sz_64k
+            )
```

Comment:
`VMALLOC_START` can't be reliably computed on aarch64. See commit torvalds/linux@9ad7c6d5e75b. I don't really have a suggestion of how to handle it though. You can see how I handle this in `drgn_tools/mm.py`. The computation here overlaps to a decent extent with that. You could make `AddrKind._ranges` public and try to reuse the value I compute (but there's no guarantee it's fully correct, just "good enough").

Or, you could simply omit it, and don't provide the VmallocTotal field on aarch64, which seems fine to me.

Reply:
I kept it calculated this way because this field has been checked and passed all vmcore tests, which means the parsed value matched exactly with the output from /proc/meminfo.

With your input, I agree that the vm layout may be changed by kernel commits. Plus, since you've already had a function to calculate address regions for vms, I'd just omit the field for aarch64.